### ISSUE TITLE
Life360 status when device offline

### ISSUE THREAD

**Issue by pnbruckner** [pnbruckner/homeassistant-config#20]
If a device being tracked by the Life360 device tracker platform goes offline -- e.g., is turned off or airplane mode is turned on -- currently the status of the device_tracker entity does not change until the device comes back online. It might be nice, if possible, for the platform to detect the device is offline and change the corresponding device_tracker entity in some way to indicate this. E.g., it might set the state to 'offline' so the device will not appear to be 'home', in another zone (if it was), etc.

**Comment by pnbruckner** [comment]
A possible way to avoid a breaking change: Add a new attribute named offline that indicates if the device is determined to be offline. Then add a new set_as_state option of offline, that will set the state to 'offline' if offline is true. This would take precedence over the other show_as_state options.

**Comment by hvbhome** [comment]
I think that would indeed be a good solution. 
FYI, I have another circle member thats currently in status: 'Lost connection', because the permissions on the device are not set to 'always' for back ground location tracking...

**Comment by pnbruckner** [comment]
I did a test. I put my phone into airplane mode. The query results from the Life360 server never had any indication that the phone was offline, other than the location information (& timestamp) becoming stale. And, BTW, during this test the Life360 website also did not indicate anything about the device being offline. The last update just kept getting older.

Also, the device_tracker component code has no API for indicating a device is offline. It only has an API to indicate the device has been "seen." And going offline is the opposite of that.

So basically there's no supported way for a device tracker platform to indicate to the device tracker component code that a device is offline. And there's no indication from the Life360 server that a device is offline. (Yes, eventually there may be errors for that device, but my test lasted over an hour and there were no errors. So even if there are errors, they are not very prompt.)

For now I think the best you can do is to use the device_tracker.life360_update_overdue event to do whatever you think appropriate.

I'll leave this issue open for now in case anyone has more input or better ideas. But for now, it doesn't look like there's much the Life360 platform code can do.

**Comment by hvbhome** [comment]
HI,

thanks for your report back.
Wouldn't it be possible to have the state show 'Overdue' when it is. Right now, the state shows the last location even when in overdue  situation. If that would change to Overdue, it would allow automations based on the state of this device_tracker more reliable?

The overdue automations in your config documentation wouldn't even be necessary then?

**Comment by pnbruckner** [comment]
I suppose it _could_ do that, but the real question is, _should_ it do that?

I've analyzed the device tracker component code quite a bit, and I'm still convinced that the only API available to the platform code is the "see" function. And that should only be used to report that the device has been seen, not that it hasn't. It's up to the component level code to decide what should happen if the device has not been seen for too long. And it does have some code for that, but it's only used with non-GPS-based devices.

It's possible for the platform code to call the see function and provide a location_name of 'Overdue'. But then what do do about the other data? I think it would have to provide the stale data, because if it sent None instead, then that would get into the "stale" processing of the component code, and that could make a device that was not home temporarily look like it _was_ home -- which obviously is not a good thing.

The other thing that making the state 'Overdue' would do is to potentially be a breaking change. Existing automations might break if all of a sudden device_tracker states went to 'Overdue', even though the device/person might actually still be home.

I understand the desire to better handle the situation when device updates get stale (whether it's because of comm errors, or just device apps & cloud servers not updating for some reason.) But I think this needs to be handled at the device tracker component level. Or, at least, that code needs to make provisions for it.

In the meantime, besides the existing events that the platform can fire, automations could also use the last_seen attribute directly.

**Comment by pnbruckner** [comment]
This issue and #21 are really two different scenarios of the same issue: what to do with the lack of new information.

**Comment by hvbhome** [comment]
Hi
choose to respond here, and not on community, or other issues raised because of your response above:

> It's possible for the platform code to call the see function and provide a location_name of 'Overdue'. But then what do do about the other data? I think it would have to provide the stale data, because if it sent None instead, then that would get into the "stale" processing of the component code, and that could make a device that was not home temporarily look like it was home -- which obviously is not a good thing.

This is exactly what is happening at the moment, and the reason I asked you in the first place... The only option I would believe would be, without changing anything in the internal logic or creating added attributes or what have you, is to show the Overdue somewhere in the attributes, or preferably state. 

Reasoning: if the system itself calculates its overdue, it cant possibly be sure of its location, so it shouldn't show a location at all, but display its uncertainty, by showing Overdue.

> The other thing that making the state 'Overdue' would do is to potentially be a breaking change. Existing automations might break if all of a sudden device_tracker states went to 'Overdue', even though the device/person might actually still be home.

I realize that, but, consider this: Better have a breaking change, and announce it (HA team never stops doing so on each update.....) and have a better, more certain component. Especially if, as you say yourself here, the device 'might' be home. In our use case, Is driving while reported to be home....

Would be  easy enough to adapt automations for a setting of Overdue?

> But I think this needs to be handled at the device tracker component level. Or, at least, that code needs to make provisions for it.

Wouldnt/can't argue with that. Since you made the device_tracker too, you'd know best if that is a better option? Glas to help and test if you'd give it a shot, goes without saying!

> automations could also use the last_seen attribute directly.

Id have to get my mind out of the knot again to see how that should be done.. 

Hope this is of any help to you, anyways, thanks for all your efforts! Much appreciated.

**Comment by pnbruckner** [comment]
>This is exactly what is happening at the moment, and the reason I asked you in the first place.

Maybe we're misunderstanding each other, but currently the code does _not_ call the see() method if the data timestamp just retrieved from the server is not newer than it was the last time it called see(). The see() method is what sends new data to the device tracker component level code. So, no, it's not sending stale data to the component level code currently.

>The only option I would believe would be, without changing anything in the internal logic or creating added attributes or what have you, is to show the Overdue somewhere in the attributes, or preferably state.

"Overdue" is effectively already being shown in the attributes in last_seen. Well, not exactly, because it's not showing it as a True/False depending on if now - last_seen > max_update_wait. But template code could calculate that (using the same value of max_update_wait.)

So I could easily add an overdue attribute if that would be generally thought useful, even if somewhat redundant. FWIW, I originally implemented the overdue/restored event feature so a person could take action, such as telling the person with the overdue device, "hey, turn your wifi back on!" or something. I wasn't considering it something useful for general presence automations. And given the reasons I've already documented, I still don't.

Which leads to whether or not it should be a possible state. Again, I really don't think it should. But, if it was, I would do it as another option for show_as_state. At least that way the individual user could decide if they want that, and it wouldn't be a breaking change. But, I still think it's trying to derive information from the lack of information, which I don't think is a good idea.

>Reasoning: if the system itself calculates its overdue, it cant possibly be sure of its location, so it shouldn't show a location at all, but display its uncertainty, by showing Overdue.

I completely get what you're saying. But that's not how any other GPS-based device tracker that I've reviewed works, and that's now how the device tracker component code treats GPS-based devices. The platform is never "sure" of a device's location until it gets new data from its ultimate source (the Life360 server in this case), and immediately it's old. That's just the nature of polling systems, which HA mostly is. BTW, the "overdue" determination is made by the user, in how they set the max_update_wait value which, also BTW, is optional.

Again, the bottom line to me is, the data is always stale. It's just a question of when does stale turn into bad/wrong/unbelievable, and what to do at that point. I'm sorry if I'm being stubborn, but so far I just don't think that the platform code should make that determination. It does already give hints, but I'm not sure it should do more than that.

BTW, I do appreciate you taking the time to argue your perspective!

**Comment by hvbhome** [comment]
cool, and i understand completely your somewhat ... reluctant. Still your suggestion:

> if it was, I would do it as another option for show_as_state. At least that way the individual user could decide if they want that, and it wouldn't be a breaking change. 

would be the preferred solution, if any. Maybe sleep another night on it, and give it a shot? ;-) Would be rrrreally helpful.

> the bottom line to me is, the data is always stale. It's just a question of when does stale turn into bad/wrong/unbelievable, and what to do at that point.

Up to know, I find the data to be very precise, and, waited and watched my wife arriving home a the east moment the Life360 tracker showed her on the map! So cool. I think we can safely say stale turns into unreliable (would be my choice of words) when threshold Overdue is met.

Thanks, and have a great one!

**Comment by pnbruckner** [comment]
First, I wholeheartedly agree, Life360 is normally very good at providing prompt updates while minimizing battery usage. I haven't found anything better. It is indeed extremely cool!

The problem, however, is that the only API a device tracker platform has is the see() method. And calling that is telling the device tracker component that the device has been seen - which it hasn't.

Consider this scenario. A device is tracked by both Life360 and a network-based platform. The network-based platform just "saw" the device, making the device_tracker entity's state 'home'. Then, right after, the Life360 platform decides the update is overdue. The only thing it can do is call see() with a location_name of 'overdue'. That changes the device_tracker entity's state from 'home' (which it is) to 'overdue', which is definitely _not_ helpful. And this is just one scenario where this doesn't work. I've thought of others, too.

There's just currently no way for platform level code to take positive action in this scenario, except for using a "side channel", which the Life360 platform is already doing by firing an event. User automations can use that to adjust the state of the overall system accordingly (other than just sending an e-mail, for example.)

**Comment by hvbhome** [comment]
> The only thing it can do is call see() with a location_name of 'overdue'. That changes the device_tracker entity's state from 'home' (which it is) to 'overdue', which is definitely not helpful.

Doesnt this only account for the life360 device_tracker? Being overdue on this tracker doesnt change the network-based tracker in any way?
I believe with this you've rather eloquently described why the overdue would have to be set... There's simply no way the device_tracker can be sure it is home, so it should not say so... (I have day to day morning experience here...)
Especially when it 'knows' it is overdue, which is exactly what this cool device_tracker sets apart from all the others. All others are in limbo, while Life360 'confesses' it doesn't know.
We really should be able to use that. I believe it it part of its magic, and creates possibilities  for new logic we aren't used to yet. 

I must yet try and use the side channel, which is on my todo list for today. Wont have it send me an email, but will try to get the overdue somewhere in the logic, so it doesn't cause a false positive state.

Woud you have any suggestions for doing so? (thinking of your composite device_tracker, which I have using the life360 device_tracker.)

**Comment by pnbruckner** [comment]
> Doesnt this only account for the life360 device_tracker? Being overdue on this tracker doesnt change the network-based tracker in any way?

Sorry, I wasn't very explicit, but I was describing the scenario where the Life360 tracker is combined with a network tracker by the component level code by the user adding the network tracker's MAC address to the Life360 tracker in known_devices.yaml. That is a feature of the tracker component that people do use. In this case the two trackers are no longer independent. The one, combined tracker is updated by both.

> I believe with this you've rather eloquently described why the overdue would have to be set... There's simply no way the device_tracker can be sure it is home, so it should not say so.

Given the current component level implementation, the only thing the platform can "say" is the device was just seen, which it does when it sees new data from the server. It has no way to say anything else. So it _doesn't_ say the device is _still seen_ between updates from the server, and it shouldn't. Sorry, but I'm not sure how to say it any other way. If it calls see() with a location_name of 'overdue' (which would indeed become the state), it's also saying at the same time, "hey, I just saw the device, and it's at the location named 'overdue'." But that's not true, and furthermore it messes up the tracker combination logic of the component level code as I've described.

FWIW, I've considered adding an overdue attribute, but the problem with that is, in order to change that from False to True, the see() method has to be called to do it. And that would effectively be the same problem as changing the state; i.e., calling see() says the device was just seen, which it wasn't. So even that wouldn't work.

I think we're just going to have to agree to disagree. Nothing I've heard (from anyone) has convinced me that my view is incorrect. The platform code just can't say the device hasn't been seen for a while without at the same time saying it was just seen. If the component code adds a feature for that I'd be happy to support it. But for now, I think it's time to close this discussion. Thanks to all who've shared their thoughts on this topic, either here or via the community forum!
